AX: Implement nameFrom: heading

rahimabdi opened the pull request:
#### 6eb2b6af4dd8b9c0bd8810a7f6c1dd7e9dc24583
<pre>
AX: Implement nameFrom: heading
<a href="https://bugs.webkit.org/show_bug.cgi?id=257186">https://bugs.webkit.org/show_bug.cgi?id=257186</a>
<a href="https://rdar.apple.com/109700680">rdar://109700680</a>

Reviewed by Tyler Wilcock.

This PR adds a nameFrom: heading fallback for determining the accessible name of
dialog, alert dialog and article elements. In the absence of a name that has been
provided by an author or by content, a name may be derived for these elements using
a depth-first search for a descendant heading element node
(using the existing DOM descendantsOfType&lt;&gt; iterator). This PR also adds an additional
AccessibilityTextSource::Heading for label type categorization and to facilitate accessibility logging.

* LayoutTests/imported/w3c/web-platform-tests/accname/name/comp_name_from_heading.tentative-expected.txt: Updated due to PR.
* LayoutTests/platform/ios/accessibility/dialog-slotted-content-expected.txt: Updated due to new test expectation.
Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::isVisibleText):
(WebCore::AXCoreObject::isDescriptiveText):
(WebCore::AXCoreObject::descriptionAttributeValue() const):
(WebCore::AXCoreObject::titleAttributeValue() const):
Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::AccessibilityTextSource):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::AXLogger::TextStream&amp; operator):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::alternativeText const):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::accessibleNameDerivesFromHeading() const):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::accessibleNameDerivesFromHeading() const):

Canonical link: <a href="https://commits.webkit.org/294335@main">https://commits.webkit.org/294335@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4c2625468f22b4b88a408a135d29a40c38e3d2f3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101464 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21130 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11434 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106617 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52093 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103504 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21437 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29623 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77272 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34302 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104471 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16539 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91630 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57614 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16360 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9649 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51441 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86229 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9721 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108969 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28593 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21027 "Passed tests") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86243 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; 33 flakes 98 failures; Uploaded test results; 14 flakes 76 failures; Compiled WebKit; 2 flakes 73 failures; Found 1 new test failure: imported/w3c/web-platform-tests/mixed-content/gen/top.http-rp/opt-in/video-tag.https.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28955 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87837 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85809 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21834 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30540 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8251 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/22711 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28524 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33804 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28335 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31655 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29894 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -346,6 +346,20 @@ bool AccessibilityObject::accessibleNameDerivesFromContent() const
     return true;
 }
 
+// If a name cannot be derived from content or author, name can be derived from the first descendant element node with a heading role.
```
Is there a spec link we can add as a comment that backs this statement up?

rahimabdi replied:
Added.

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -346,6 +346,20 @@ bool AccessibilityObject::accessibleNameDerivesFromContent() const
+// If a name cannot be derived from content or author, name can be derived from the first descendant element node with a heading role.
+bool AccessibilityObject::accessibleNameDerivesFromHeading() const
+{
+    // Only alert dialog, dialog and article elements support nameFrom: heading.
```
Is there a spec link we can add as a comment that backs this statement up?

rahimabdi replied:
Removed this and spec link added in the above comment (also clarified in the comment above "on permitted roles".)

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1819,6 +1820,24 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
 
+    // Dialog, alert dialog and article elements may derive their name from the first descendant node
+    // that is a heading (depth-first search, preorder traversal).
+    if (AccessibilityObject::accessibleNameDerivesFromHeading()) {
```
Is the `AccessibilityObject::` prefix necessary?

rahimabdi replied:
Nope, removed.

The method `accessibleNameDerivesFromHeading()` is defined in `AccessibilityObject`, and the derived `AccessibilityNodeObject` class object should be able to call it out without scope resolution since it's not overrode (this note is for myself 😄).

twilco replied:
Seems like it's still present in the latest commit 🤔 also, even it were overridden by `AccessibilityNodeObject`, you'd be able to (and in fact _want to_) call it without prefix to ensure the most-subclass implementation was called.

rahimabdi replied:
Apologies, removed.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1819,6 +1820,24 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
+    // Dialog, alert dialog and article elements may derive their name from the first descendant node
+    // that is a heading (depth-first search, preorder traversal).
+    if (AccessibilityObject::accessibleNameDerivesFromHeading()) {
+        if (auto* rootNode = dynamicDowncast<ContainerNode>(node)) {
```
Maybe `containerNode` instead of `rootNode`? "Root" has strong meanings about the top-level object in the tree, which I don't think is the intention here.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1819,6 +1820,24 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
         textOrder.append(AccessibilityText(accessibleNameForNode(*node), AccessibilityTextSource::Alternative));
 
+    // Dialog, alert dialog and article elements may derive their name from the first descendant node
+    // that is a heading (depth-first search, preorder traversal).
```
I think it could be a bit cleaner to have this comment inside the `if () { }`. Also, maybe we shouldn't call out the specific roles here? As otherwise this comment could get stale if the implementation of `accessibleNameDerivesFromHeading` changes. Maybe there's a more general way we can write it.

rahimabdi replied:
Agreed, although I can't think of a descriptor that describes these 3 specific elements. What do you think of:

```
// Where an element supports nameFrom: heading and no nameFrom: content/author is supplied, its accname may be derived 
// from the first descendant node that is a heading (depth-first search, preorder traversal).
```

twilco replied:
Excellent!

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1819,6 +1820,24 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
+                        if (!nameFromHeading.isEmpty())
+                            textOrder.append(AccessibilityText(nameFromHeading, AccessibilityTextSource::Heading));
+                    }
+                }
```
Few things here:

1. Do we need to cast to `AccessibilityNodeObject`? AFAICT all the methods we call on this object are available on the base `AXCoreObject` class, but maybe I'm missing one.
2. We do need to null-check `axObjectCache()` before using it (even though many existing places don't, it's bad practice and has caused crashes in the past)

One way to write this whole section would be:

```cpp
if (AccessibilityObject::accessibleNameDerivesFromHeading()) {
    CheckedPtr cache = axObjectCache();
    // Dialog, alert dialog and article elements may derive their name from the first descendant node
    // that is a heading (depth-first search, preorder traversal).
    if (auto* containerNode = dynamicDowncast<ContainerNode>(node); containerNode && cache) {
        for (auto& element : descendantsOfType<Element>(*containerNode)) {
            if (auto* descendantObject = cache->getOrCreate(element); descendantObject && descendantObject->isHeading()) {
                TextUnderElementMode mode;
                mode.includeFocusableContent = true;
                String nameFromHeading = descendantObject->textUnderElement(mode);
                if (!nameFromHeading.isEmpty())
                    textOrder.append(AccessibilityText(nameFromHeading, AccessibilityTextSource::Heading));
            }
        }
    }
}
```

rahimabdi replied:
> Do we need to cast to AccessibilityNodeObject? AFAICT all the methods we call on this object are available on the base AXCoreObject class, but maybe I'm missing one.

Good catch, removed.

> We do need to null-check axObjectCache() before using it (even though many existing places don't, it's bad practice and has caused crashes in the past)

Got it. The rewrite is more concise/elegant, thank you!

twilco reviewed (commented):
Almost there! Left one more comment, and also we try to include the Radar link in the commit message and PR description if you wouldn't mind changing that too 🙏

rahimabdi committed:
AX: Implement nameFrom: heading
https://bugs.webkit.org/show_bug.cgi?id=257186
rdar://109700680

Reviewed by Tyler Wilcock.

This PR adds a nameFrom: heading fallback for determining the accessible name of
dialog, alert dialog and article elements. In the absence of a name that has been
provided by an author or by content, a name may be derived for these elements using
a depth-first search for a descendant heading element node
(using the existing DOM descendantsOfType<> iterator). This PR also adds an additional
AccessibilityTextSource::Heading for label type categorization and to facilitate accessibility logging.

* LayoutTests/imported/w3c/web-platform-tests/accname/name/comp_name_from_heading.tentative-expected.txt: Updated due to PR.
* LayoutTests/platform/ios/accessibility/dialog-slotted-content-expected.txt: Updated due to new test expectation.
Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::isVisibleText):
(WebCore::AXCoreObject::isDescriptiveText):
(WebCore::AXCoreObject::descriptionAttributeValue() const):
(WebCore::AXCoreObject::titleAttributeValue() const):
Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::AccessibilityTextSource):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::AXLogger::TextStream& operator):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::alternativeText const):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::accessibleNameDerivesFromHeading() const):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::accessibleNameDerivesFromHeading() const):

Canonical link: https://commits.webkit.org/294335@main
